source/extensions/stat_sinks/metrics_service/grpc_metrics_service_impl.cc

```diff
@@ -139,6 +133,28 @@ void MetricsFlusher::flushHistogram(io::prometheus::client::MetricFamily& summar
     bucket->set_cumulative_count(hist_stats.computedBuckets()[i]);
   }
 }
+
+io::prometheus::client::Metric*
+MetricsFlusher::populateMetricsFamily(io::prometheus::client::MetricFamily& metrics_family,
+                                      io::prometheus::client::MetricType type,
+                                      int64_t snapshot_time_ms, const Stats::Metric& metric) const {
+  metrics_family.set_type(type);
+  auto* prometheus_metric = metrics_family.add_metric();
+  prometheus_metric->set_timestamp_ms(snapshot_time_ms);
+
+  if (emit_labels_) {
+    metrics_family.set_name(metric.tagExtractedName());
+    for (const auto& tag : metric.tags()) {
```

Comment:
hmm. I'm wondering if this is going to be a perf issue as constructing the tag array is going to take symbol table locks and do some string-building, with those strings then being copied again below.

It might be slightly better to use metric::iterateTagStatNames but we'd still need to convert to strings here. Maybe this transformation should be cached somehow in the MetricsFlusher object?

Maybe leave a note here to take a look at the perf implications of this later? The reason I'm concerned about this is that it looks like this is called on every flush, each with its own value, but it's re-creating a bunch of metadata for each stat that doesn't change. It'd be nice to factor that repeated computation out.

Reply:
Yeah let me leave a note here to look into the perf of this. 

I think there are two options that could be explored: 1) cache just the symbol -> string conversion just to avoid the locks on the symbol table or 2) cache the entire prom Metric per stat. I'm not sure how much 2) gains us over 1) since we'll need to copy the proto completely anyways, so it might just result in a lot of wasted space keeping all these protos stored in memory? At least with 1) we get the benefit of space savings if a symbol is reused between stats. In any case I will leave this as a TODO for now

Reply:
Sounds good. One minor concern to note also is that if we add a cache, it could retain storage for stats that belong to deleted scopes, thus leaking. We can fix all that but it's definitely a separate PR.